File: backend/app/schemas/workflow.py
"""工作流预览模块的请求与响应模型定义"""

from datetime import datetime

from pydantic import BaseModel, Field


class WorkflowPreviewRequestSchema(BaseModel):
    """工作流预览请求模型，允许指定是否强制重新规划"""

    force_replan: bool = False
    pause_after_nodes: list[str] = []


class WorkflowControlRequestSchema(BaseModel):
    """工作流控制请求模型，描述暂停、恢复、中断或改向操作"""

    action: str
    redirect_instruction: str = ""


class RequirementSummarySchema(BaseModel):
    """结构化需求摘要模型，描述目标、约束与产出类型"""

    goal: str
    constraints: list[str]
    output_types: list[str]
    context: str


class WorkflowNodeSchema(BaseModel):
    """工作流节点模型，描述角色、任务和依赖关系"""

    id: str
    template_id: str
    template_source: str = "builtin"
    template_summary: str
    role: str
    task: str
    tools: list[str]
    llm: str
    max_retries: int = Field(ge=0)
    depends_on: list[str]
    trigger_keywords: list[str] = []
    runtime_status: str | None = None


class WorkflowDagSchema(BaseModel):
    """工作流 DAG 模型，描述节点列表与执行模式"""

    nodes: list[WorkflowNodeSchema]
    execution_mode: str


class WorkflowExecutionLogSchema(BaseModel):
    """工作流执行日志模型，描述节点完成后的最小交接摘要"""

    node_id: str
    role: str
    summary: str
    status: str
    artifacts: list[str] = []


class WorkflowHandoffSchema(BaseModel):
    """工作流交接记录模型，描述节点之间的共享工作区交付信息"""

    from_agent: str
    to_agent: str
    summary: str
    artifacts: list[str]
    created_at: str


class WorkflowWorkspaceStateSchema(BaseModel):
    """工作流共享工作区状态模型，描述当前目录、进度和产出物快照"""

    workspace_path: str
    status: str
    progress: int
    active_node_id: str | None
    artifacts: list[str]
    updated_at: str
    pause_after_nodes: list[str] = []


class WorkflowProjectMemoryErrorSchema(BaseModel):
    """项目级记忆中的最近错误摘要"""

    workflow_id: int
    node_id: str
    role: str
    error_message: str
    recovery_suggestion: str
    updated_at: str


class WorkflowProjectMemorySchema(BaseModel):
    """项目级记忆响应模型，描述对话级共享长期上下文"""

    memory_path: str
    latest_goal: str
    active_constraints: list[str]
    key_points: list[str]
    artifacts: list[str]
    workflow_count: int
    latest_error: WorkflowProjectMemoryErrorSchema | None = None
    updated_at: str


class WorkflowArtifactSchema(BaseModel):
    """工作流产物描述模型，供前端按类型渲染预览组件"""

    name: str
    relative_path: str
    preview_type: str
    mime_type: str
    size_bytes: int
    updated_at: str


class WorkflowRunSchema(BaseModel):
    """工作流运行状态模型，描述当前轮次与断点控制信息"""

    run_id: int | None
    status: str
    control_signal: str
    checkpoint_node_id: str | None
    redirect_instruction: str
    saved_at: str | None


class WorkflowErrorReportSchema(BaseModel):
    """工作流错误报告模型，描述失败节点、回滚信息与恢复建议"""

    failed_node_id: str
    failed_role: str
    task: str
    error_message: str
    retry_count: int
    max_retries: int
    can_retry: bool
    upstream_node_id: str | None
    upstream_role: str
    rollback_checkpoint_node_id: str | None
    rollback_progress: int
    recovery_suggestion: str | None = None


class WorkflowPreviewResponseSchema(BaseModel):
    """工作流预览响应模型，返回需求摘要和 DAG 结果"""

    workflow_id: int
    conversation_id: int
    status: str
    progress: int
    requirement: RequirementSummarySchema
    dag: WorkflowDagSchema
    execution_logs: list[WorkflowExecutionLogSchema]
    workspace: WorkflowWorkspaceStateSchema
    project_memory: WorkflowProjectMemorySchema
    handoff_logs: list[WorkflowHandoffSchema]
    workflow_run: WorkflowRunSchema
    error_report: WorkflowErrorReportSchema | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/workflow.ts
// 工作流预览请求类型
export interface WorkflowPreviewRequest {
  force_replan: boolean
  pause_after_nodes: string[]
}

// 工作流控制请求类型
export interface WorkflowControlRequest {
  action: 'pause' | 'resume' | 'abort' | 'redirect'
  redirect_instruction?: string
}

// 结构化需求摘要类型
export interface RequirementSummary {
  goal: string
  constraints: string[]
  output_types: string[]
  context: string
}

// 工作流节点类型
export interface WorkflowNode {
  id: string
  template_id: string
  template_source: string
  template_summary: string
  role: string
  task: string
  tools: string[]
  llm: string
  max_retries: number
  depends_on: string[]
  trigger_keywords: string[]
  runtime_status?: string
}

// 工作流 DAG 类型
export interface WorkflowDag {
  nodes: WorkflowNode[]
  execution_mode: string
}

// 工作流执行日志类型
export interface WorkflowExecutionLog {
  node_id: string
  role: string
  summary: string
  status: string
  artifacts: string[]
}

// 工作流产物预览类型
export type WorkflowArtifactPreviewType = 'code' | 'image' | 'document' | 'binary'

// 工作流产物描述类型
export interface WorkflowArtifact {
  name: string
  relative_path: string
  preview_type: WorkflowArtifactPreviewType
  mime_type: string
  size_bytes: number
  updated_at: string
}

// 工作流交接记录类型
export interface WorkflowHandoffLog {
  from_agent: string
  to_agent: string
  summary: string
  artifacts: string[]
  created_at: string
}

// 工作流共享工作区状态类型
export interface WorkflowWorkspaceState {
  workspace_path: string
  status: string
  progress: number
  active_node_id: string | null
  artifacts: string[]
  updated_at: string
  pause_after_nodes: string[]
}

// 项目级记忆中的最近错误摘要
export interface WorkflowProjectMemoryError {
  workflow_id: number
  node_id: string
  role: string
  error_message: string
  recovery_suggestion: string
  updated_at: string
}

// 项目级记忆类型
export interface WorkflowProjectMemory {
  memory_path: string
  latest_goal: string
  active_constraints: string[]
  key_points: string[]
  artifacts: string[]
  workflow_count: number
  latest_error: WorkflowProjectMemoryError | null
  updated_at: string
}

// 工作流运行状态类型
export interface WorkflowRunState {
  run_id: number | null
  status: string
  control_signal: string
  checkpoint_node_id: string | null
  redirect_instruction: string
  saved_at: string | null
}

// 工作流错误恢复报告类型
export interface WorkflowErrorReport {
  failed_node_id: string
  failed_role: string
  task: string
  error_message: string
  retry_count: number
  max_retries: number
  can_retry: boolean
  upstream_node_id: string | null
  upstream_role: string
  rollback_checkpoint_node_id: string | null
  rollback_progress: number
  recovery_suggestion: string | null
}

// 工作流预览响应类型
export interface WorkflowPreview {
  workflow_id: number
  conversation_id: number
  status: string
  progress: number
  requirement: RequirementSummary
  dag: WorkflowDag
  execution_logs: WorkflowExecutionLog[]
  workspace: WorkflowWorkspaceState
  project_memory: WorkflowProjectMemory
  handoff_logs: WorkflowHandoffLog[]
  workflow_run: WorkflowRunState
  error_report: WorkflowErrorReport | null
  created_at: string
  updated_at: string
}
